cpp/src/dbscan/vertexdeg/precomputed.cuh

```diff
@@ -0,0 +1,142 @@
+  uint32_t sid = local_write_i * block_rows + local_write_j;
+
+  __syncthreads();
+
+  // Write tile in adj
+  if (write_j < N && write_i < batch_size)
+    adj[write_j * batch_size + write_i] = shared_mem[sid];
+}
+
+/**
+ * Calculates the vertex degree array and the epsilon neighborhood adjacency matrix for the batch.
+ */
+template <typename value_t, typename index_t = int>
+void launcher(const raft::handle_t& handle, Pack<value_t, index_t> data,
+              index_t start_vertex_id, index_t batch_size,
+              cudaStream_t stream) {
+  const value_t& eps = data.eps;
+
+  /// TODO: case where less remaining elements than batch_size?
+
+  // Note: the matrix is symmetric. We take advantage of this to have two
+  //       coalesced kernels:
+  //  - The reduction works on a column-major N*B matrix to compute a B vector
+  //    with a cub-BlockReduce-based primitive.
+  //    The final_op is used to compute the total number of non-zero elements
+  //  - The conversion to a boolean matrix works on a column-major B*N matrix
+  //    (coalesced 2d copy + transform).
+
+  // Regarding index types, a special index type is used here for indices in
+  // the distance matrix due to its dimensions (that are independent of the
+  // batch size)
+  using long_index_t = long long int;
+
+  // Reduction to compute the vertex degrees
+  index_t* d_nnz = data.vd + batch_size;
+  CUDA_CHECK(cudaMemsetAsync(d_nnz, 0, sizeof(index_t), stream));
+  raft::linalg::coalescedReduction<value_t, index_t, long_index_t>(
+    data.vd, data.x + start_vertex_id * data.N, data.N, batch_size, (index_t)0,
+    stream, false,
+    [eps] __device__(value_t dist, long_index_t idx) {
+      return static_cast<index_t>(dist <= eps);
+    },
+    raft::Sum<index_t>(),
+    [d_nnz] __device__(index_t degree) {
+      atomicAdd(d_nnz, degree);
+      return degree;
+    });
+
+  // Transform the distance matrix into a neighborhood matrix
+  dist_to_adj_kernel<<<data.N, std::min(batch_size, (index_t)256), 0, stream>>>(
+    data.x, data.adj, (long_index_t)data.N, (long_index_t)start_vertex_id,
+    (long_index_t)batch_size, data.eps);
+  CUDA_CHECK(cudaPeekAtLastError());
+
+  // Note: in case we want to use the same N*B sub-matrix to compute adj,
+  // e.g for multi-node with partitioned matrix, we can use the alternative
+  // kernel that transposes with shared memory (commented code below)
+
+  // constexpr uint32_t block_rows = 32, block_cols = 32;
+  // dim3 grid(raft::ceildiv<index_t>(data.N, block_rows),
+  //           raft::ceildiv<index_t>(batch_size, block_cols));
+  // dim3 block(block_rows * block_cols);
+  // dist_to_adj_transposed_kernel<block_rows, block_cols>
+  //   <<<grid, block, 0, stream>>>(data.x, data.adj, (long_index_t)data.N,
+  //                                (long_index_t)start_vertex_id,
+  //                                (long_index_t)batch_size, data.eps);
+  // CUDA_CHECK(cudaPeekAtLastError());
```

Comment:
Remove this block, and update the comment above to explain what needs to be handled differently if the data is distributed.